src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
+    {
+        private class ConnectionPools
+        {
+            private readonly HubClient _hubClient;
+            private readonly HostGroup _hostGroup;
+            private readonly TimeSpan _timeout;
+
+            private readonly ReaderWriterLockSlim _shutdownLock;
+            private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
+
+            private readonly int _maxPoolConnections;
+            private readonly ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>> _pools;
+
+            // indicate whether pool should be used.
+            // it is mutable since it will set to false when this pool got shutdown
+            private volatile bool _usePool;
```

Comment:
this seems strange.  if needed, why not have the bool record what is actually the state change.  i.e. "_isShutdown".  then you can decide if the pool is used based on that.

Reply:
Another way to do this sort of thing is to have whoever *owns* the pool, swap it out with a no-op version upon shutdown.

Reply:
I don't want RemoteHostClient is disruptive. if a feature haven't had change to move to new client yet. I don't want it to suddenly fail. it will still work, just not use pool.

Reply:
I thought out shutdown but usePool seems more appreciate. since shutdown cause pool not to be used. but CreateConnection will still work if someone chose to still use this RemostHostClient.

Reply:
Please document all this in the class :)

Reply:
sure

Reply:
💭 I agree this is super strange. To me it doesn't make sense for the connection pool to be creating non-pooled connections at all.